### PR TITLE
feat: Component Mixin to Receive Cursor Hover Events

### LINKED ISSUE #1818: Add cursor components mixin to Game
# Problem to solve

It would be convenient to propagate PointerHoverEvents to child Components, not just handle them from the base game.

# Proposal

Make a HasCursorHandlerComponents mixin on FlameGame, could be very similar to HasKeyboardHandlerComponents.  See example below:
```dart
mixin HasCursorHandlerComponents on FlameGame implements MouseMovementDetector {
  @override
  @mustCallSuper
  void onMouseMove(PointerHoverInfo info) {
    propagateToChildren<CursorHandler>(
      (CursorHandler child) => child.onMouseMove(info),
    );
  }
}
```
# More information

Interested in a PR, basic implementation already tested.

### PR DESCRIPTION
**Pull request by garysm**
# Description

This PR adds the ```HasCursorHandlerComponents``` mixin on FlameGame, and the ```CursorHandler``` mixin on ```Component```.  Heavily inspired by ```HasKeyboardHandlerComponents```.

```HasCursorHandlerComponents``` propagates cursor hover events to child ```CursorHandler``` ```Components```, making these events easier to handle.

## Checklist

- [x] The title of my PR starts with a [Conventional Commit] prefix (`fix:`, `feat:`, `docs:` etc).
- [x] I have read the [Contributor Guide] and followed the process outlined for submitting PRs.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples`.

## Breaking Change

- [ ] Yes, this is a breaking change.
- [x] No, this is *not* a breaking change.

## Related Issues

Closes #1818 

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md
[Conventional Commit]: https://conventionalcommits.org

### CODE REVIEW

File: packages/flame/lib/src/gestures/detectors.dart [hunk 1]
```diff
@@ -191,7 +192,14 @@ mixin ScaleDetector on Game {
 }
 
 mixin MouseMovementDetector on Game {
-  void onMouseMove(PointerHoverInfo info) {}
+  void onMouseMove(PointerHoverInfo info) {
+    assert(
+      this is! HasCursorHandlerComponents,
+      'A mouse movement event was registered by MouseMovementDetector for a game also '
+      'mixed with HasCursorHandlerComponents. Do not mix with both, '
+      'HasCursorHandlerComponents removes the necessity of MouseMovementDetector',
```
Comment (spydon) [inline packages/flame/lib/src/gestures/detectors.dart:200]: Analyze is failing due to these lines being too long.

**Review by spydon — COMMENTED** [review]
Lgtm except for the too long lines!

**Comment by spydon** [comment]
@garysm I fixed the too long lines, could you add some docs for the functionality?
Possibly also an example for the examples directory.

**Comment by garysm** [comment]
@spydon Yes, I will work on docs and provide an example.  Thank you for fixing the too long lines.

**Comment by alestiago** [comment]
Would we like to remove this mixin if #1733 gets done?

**Comment by garysm** [comment]
@alestiago, that would probably make sense.  I can finish up with this, then convert to a class once #1733 is implemented.

**Comment by garysm** [comment]
Added docs and example @spydon .  Let me know if you'd like a test to be written as well.

**Comment by spydon** [comment]
Good job @garysm, I'll hopefully have time to have a look in the coming days!

@st-pasha didn't you have some feedback on this PR?

**Comment by st-pasha** [comment]
Yes, I'll re-review this PR. But first we need to fix up this branch -- currently it shows 54 files changed.

**Comment by garysm** [comment]
> Yes, I'll re-review this PR. But first we need to fix up this branch -- currently it shows 54 files changed.

I performed a rebase, that is probably why.

**Comment by spydon** [comment]
> I performed a rebase, that is probably why.

GitHub should be able to understand rebases, but maybe not when the branch is outdated, can you press "Update branch" in the UI and see if one of those merges help?

**Comment by garysm** [comment]
@spydon that suggestion worked, thanks!

**Review by st-pasha — COMMENTED** [review]
Some general comments:

1. Tests are currently missing. Generally, in order to test mouse movement events in a test, you would `testWidgets()`, and then `tester.createGesture()` + `gesture.moveBy()`.

2. The interaction between this mixin and `HasDraggableComponents` is unclear: would they coexist peacefully, or will one preempt the other? This needs to be documented. Also, if both mixins are present and the user drags the mouse - would both mixins process this event?

3. We're slowly trying to move towards more uniform approach for event handling, see the `lib/src/events/` folder. I'm not sure whether you'd want to modify your PR to use this newer approach, but that would save us some work in the future.

**Comment by garysm** [comment]
Thanks for the feedback @st-pasha , I'll work on this.

**Comment by spydon** [comment]
Updated this branch with `main` so that it gets the markdown-lint step too

**Comment by spydon** [comment]
@garysm now you have some time to score this PR for Hacktoberfest. ;)

**Review by renancaraujo — APPROVED** [review]

File: doc/flame/inputs/gesture-input.md [hunk 2]
```diff
@@ -488,3 +488,37 @@ More information about how to define hitboxes can be found in the hitbox section
...
+### MouseMovement on the game level
+
+To give a `Game` class the ability to handle mouse movement/hover events, use the `MouseMovementDetector` mixin.  
+
+You can then override an `onMouseMove` method. This method receives [`PointerHoverInfo`](https://pub.dev/documentation/flame/latest/events/PointerHoverInfo-class.html) which provides the position of the hover event, as well as the raw [`PointerHoverEvent`](https://api.flutter.dev/flutter/gestures/PointerHoverEvent-class.html)
+
+
+
+> ⚠️ Note: This should not be used on platforms that do not 
+> support mouse cursors.
```
[the remark this answers is not in the document]
Comment (spydon) [reply doc/flame/inputs/gesture-input.md:503]: @garysm did you see this comment?

File: doc/flame/inputs/gesture-input.md [hunk 3]
```diff
@@ -490,3 +490,37 @@ More information about how to define hitboxes can be found in the hitbox section
...
+    print('Y: ${info.eventPosition.game.y}');
+  }
+}
+
+```
+
+
+### MouseMovement on the component level
+
+To recieve mouse movement/hover events in components, use the `HasCursorHandlerComponents` mixin on any `Game` sub class first, and remove the `MouseMovementDetector` mixin (if present).
```
Comment (spydon) [inline doc/flame/inputs/gesture-input.md:522]:
```suggestion
To receive mouse movement/hover events in components, use the `HasCursorHandlerComponents` mixin on any `Game` sub class first, and remove the `MouseMovementDetector` mixin (if present). 
```

File: packages/flame/lib/src/gestures/detectors.dart [hunk 4]
```diff
@@ -191,7 +192,15 @@ mixin ScaleDetector on Game {
...
 mixin MouseMovementDetector on Game {
-  void onMouseMove(PointerHoverInfo info) {}
+  void onMouseMove(PointerHoverInfo info) {
+    assert(
+      this is! HasCursorHandlerComponents,
+      'A mouse movement event was registered by MouseMovementDetector for a '
+      'game also mixed with HasCursorHandlerComponents. Do not mix with both, '
+      'HasCursorHandlerComponents removes the necessity of '
+      'MouseMovementDetector.',
+    );
```
[the remark this answers is not in the document]
Comment (spydon) [reply packages/flame/lib/src/gestures/detectors.dart:202]: Did you see this comment @garysm?

**Comment by garysm** [comment]
Just saw @spydon, thanks.  Hoping to finish this up as I now have more time on my hands.

File: doc/flame/inputs/gesture-input.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (garysm) [reply doc/flame/inputs/gesture-input.md:503]: Reviewing these comments now @spydon, thanks. Hoping to finish this up as I now have more time on my hands.

File: doc/flame/inputs/gesture_input.md [hunk 5]
```diff
@@ -490,3 +490,44 @@ More information about how to define hitboxes can be found in the hitbox section
...
+To give a `Game` class the ability to handle mouse movement/hover events,
+use the `MouseMovementDetector` mixin.  
+
+You can then override an `onMouseMove` method. This method receives
+[`PointerHoverInfo`](https://pub.dev/documentation/flame/latest/events/PointerHoverInfo-class.html)
+which provides the position of the hover event, as well as the raw
+[`PointerHoverEvent`](https://api.flutter.dev/flutter/gestures/PointerHoverEvent-class.html)
+
+```{warning}
+This should not be used on platforms that do not support mouse cursors.
```
Comment (luanpotter) [inline doc/flame/inputs/gesture_input.md:506]: can we make this warning more clear as to what happens?
is the handler ignored?
or will it error?
that is relevant to multi-platform games

File: examples/lib/stories/input/mouse_movement_components_example.dart [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  Future<void> onLoad() async {
+    add(
+      CursorFollowerCircle(
+        radius: 20.0,
+      )..position = size / 2,
+    );
+    add(
+      CursorFollowerCircle(
+        radius: 20.0,
+      )..position = Vector2(100.0, 100.0),
```
Comment (luanpotter) [inline examples/lib/stories/input/mouse_movement_components_example.dart:26]: Vector2.all

File: examples/lib/stories/input/mouse_movement_components_example.dart [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+  @override
+  void render(Canvas canvas) {
+    super.render(canvas);
+
+    canvas.drawCircle(
+      Offset(_radius, _radius),
+      _radius,
+      _paint,
+    );
+    canvas.drawLine(
```
Comment (luanpotter) [inline examples/lib/stories/input/mouse_movement_components_example.dart:64]: (not part of this PR)
we should have a renderLine ext similar to renderPoint
https://github.com/flame-engine/flame/blob/main/packages/flame/lib/src/extensions/canvas.dart#L21

File: examples/lib/stories/input/mouse_movement_components_example.dart (hunk 6 above)
Comment (erickzanardo) [reply examples/lib/stories/input/mouse_movement_components_example.dart:26]:
```suggestion
      )..position = Vector2.all(100),
```

**Review by erickzanardo — APPROVED** [review]

**Comment by spydon** [comment]
@garysm are you interested in continuing working on this? :)

**Comment by garysm** [comment]
> @garysm are you interested in continuing working on this? :)

Yes, has been on the backburner for me @spydon, but I need to finish this off.  I need to take a look at the event handling as @st-pasha mentioned and catch up on Flame in general, to save you all some effort :)

**Comment by garysm** [comment]
@spydon taking a look at #2450, this makes a lot more sense and aligns to what I'd like to achieve.  Since there is no existing cursor component mixin, there shouldn't be a need for a bridge either. I can re-work this PR for "new-style" components and event handling. Or I can close and create a new one for that, whatever works.

**Comment by spydon** [comment]
> @spydon taking a look at #2450, this makes a lot more sense and aligns to what I'd like to achieve.  Since there is no existing cursor component mixin, there shouldn't be a need for a bridge either. I can re-work this PR for "new-style" components and event handling. Or I can close and create a new one for that, whatever works.

Whatever is the easiest for you :)

**Comment by garysm** [comment]
It seems to me that I cannot add a hover listener directly to the game from my new `CursorHoverDispatcher` as (to my knowledge) Flutter does not have a built-in hover `GestureRecognizer`, which is required to add to the `GestureRecognizerBuilder`.  Should there be a new method created which specifically adds only mouse detectors, and isn't dependent on a mixin on `Game`?

**Comment by garysm** [comment]
@spydon any thoughts on this?

**Comment by spydon** [comment]
@garysm I missed that you had written! If it is possible (which it should be) it should be done in the same manner as the other ones, i.e. without a mixin in the game, there should be some hints on what to connect the game to in the old hover functionality.

**Comment by garysm** [comment]
> @garysm I missed that you had written! If it is possible (which it should be) it should be done in the same manner as the other ones, i.e. without a mixin in the game, there should be some hints on what to connect the game to in the old hover functionality.

I don't think that the current `GestureDetectorBuilder` is suitable to handle a `PointerHoverEvent`, as (to my knowledge) the `RawGestureDetector` returned by the builder will not accept non-contact gestures.  The solution I think would be to create a separate `MouseDetectorBuilder`, which would similarly rebuild the game widget when a new listener is added:

```dart
class MouseDetectorBuilder {
  MouseDetectorBuilder([this._onChange]);

  final Map _listeners = {};
  final void Function()? _onChange;

  Widget build(Widget child) {
    // Return mouse listeners
  }
}

```
My question then would be, what should the widget returned from this builder be? I am unsure, since the listeners cannot be passed into a RawGestureDetector, but should still be consolidated into a single widget.

What are your thoughts @spydon ?

**Comment by spydon** [comment]
> My question then would be, what should the widget returned from this builder be? I am unsure, since the listeners cannot be passed into a RawGestureDetector, but should still be consolidated into a single widget.
> 
> What are your thoughts @spydon ?

Sorry for the late reply, it's been very busy with FlutterCon!
Couldn't this be transformed to the new system: https://github.com/flame-engine/flame/blob/main/packages/flame/lib/src/game/game_widget/gesture_detector_builder.dart#L180-L194 ?

**Comment by garysm** [comment]
> > My question then would be, what should the widget returned from this builder be? I am unsure, since the listeners cannot be passed into a RawGestureDetector, but should still be consolidated into a single widget.
> > What are your thoughts @spydon ?
> 
> Sorry for the late reply, it's been very busy with FlutterCon! Couldn't this be transformed to the new system: https://github.com/flame-engine/flame/blob/main/packages/flame/lib/src/game/game_widget/gesture_detector_builder.dart#L180-L194 ?

No problem! The issue would be that there is no way to dynamically add `onMouseMove` or similar callbacks like `GestureDetectorBuilder`, as `applyMouseDetectors` still relies on the old system to apply each callback when needed.  That's why I think it might be appropriate to create a separate builder with those capabilities, but I'm not sure what that might look like.

**Comment by spydon** [comment]
> No problem! The issue would be that there is no way to dynamically add `onMouseMove` or similar callbacks like `GestureDetectorBuilder`, as `applyMouseDetectors` still relies on the old system to apply each callback when needed. That's why I think it might be appropriate to create a separate builder with those capabilities, but I'm not sure what that might look like.

Ah right, now I understand the problem.
Yeah, seems like a separate builder would be needed then, but that builder would contain pretty much everything (and some more, like in the other builder) from the snippet I sent previously I think?

**Comment by spydon** [comment]
Superseded by #2706, thanks a lot for this PR though @garysm even though not merged it gave a lot of insights!
